hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/map/MapPutAllWithMetadataMessageTask.java

```diff
@@ -0,0 +1,119 @@
+import com.hazelcast.spi.merge.SplitBrainMergeTypes.MapMergeTypes;
+
+import java.security.Permission;
+import java.util.ArrayList;
+import java.util.List;
+
+import static com.hazelcast.spi.impl.merge.MergingValueFactory.createMergingEntry;
+
+public class MapPutAllWithMetadataMessageTask
+        extends AbstractMapPartitionMessageTask<MapPutAllWithMetadataCodec.RequestParameters> {
+
+    private volatile long startTimeNanos;
+
+    public MapPutAllWithMetadataMessageTask(ClientMessage clientMessage, Node node, Connection connection) {
+        super(clientMessage, node, connection);
+    }
+
+    @Override
+    protected Operation prepareOperation() {
+        SerializationService ss = nodeEngine.getSerializationService();
+
+        List<MapMergeTypes<Object, Object>> mergingEntries = new ArrayList<>();
+        for (SimpleEntryView<Data, Data> entry : parameters.entries) {
+            mergingEntries.add(createMergingEntry(ss, entry));
+        }
+
+        SplitBrainMergePolicy mergePolicy = nodeEngine.getSplitBrainMergePolicyProvider()
+                                                      .getMergePolicy(PassThroughMergePolicy.class.getName());
+
+        return new MergeOperation(parameters.name, mergingEntries, mergePolicy, false);
+    }
+
+    @Override
+    protected MapPutAllWithMetadataCodec.RequestParameters decodeClientMessage(ClientMessage clientMessage) {
+        return MapPutAllWithMetadataCodec.decodeRequest(clientMessage);
+    }
+
+    @Override
+    protected ClientMessage encodeResponse(Object response) {
+        return MapPutAllWithMetadataCodec.encodeResponse();
+    }
+
+    @Override
+    public String getServiceName() {
+        return MapService.SERVICE_NAME;
+    }
+
+    @Override
+    protected void beforeProcess() {
+        startTimeNanos = Timer.nanos();
+    }
+
+    @Override
+    protected Object processResponseBeforeSending(Object response) {
+        MapService mapService = getService(MapService.SERVICE_NAME);
+        MapContainer mapContainer = mapService.getMapServiceContext().getMapContainer(parameters.name);
+        if (mapContainer.getMapConfig().isStatisticsEnabled()) {
+            mapService.getMapServiceContext().getLocalMapStatsProvider().getLocalMapStatsImpl(parameters.name)
+                    .incrementPutLatencyNanos(parameters.entries.size(),
+                            Timer.nanosElapsed(startTimeNanos));
+        }
+        return response;
+    }
+
+    @Override
+    public Permission getRequiredPermission() {
+        return new MapPermission(parameters.name, ActionConstants.ACTION_PUT);
+    }
+
+    @Override
+    public String getDistributedObjectName() {
+        return parameters.name;
+    }
+
+    @Override
+    public String getMethodName() {
+        return "putAllWithMetadata";
+    }
+
+    @Override
+    public Object[] getParameters() {
+        List<SimpleEntryView<Data, Data>> entries = new ArrayList<>(parameters.entries);
+        return new Object[]{entries};
```

Comment:
I don't see any strong contract for this. The following doesn't provide many details:

https://github.com/hazelcast/hazelcast/blob/7c3894b1474dba56648cf10ad8441615486fcdea/hazelcast/src/main/java/com/hazelcast/client/impl/client/SecureRequest.java#L46-L53

`putAll` takes `Map<K, V>` so returning `Map<Data, Data>` makes sense, but `putAllWithMetadataAsync` takes `Collection<EntryView<K,V>>`, so that's why I returned a list here.

Reply:
ok, makes sense.